server/plugin_release.go

```diff
@@ -750,53 +751,57 @@ func getPluginSigningSftpClient(cfg *MatterbuildConfig) (*sftp.Client, error) {
 	return sftp, nil
 }
 
-// hasExpectedPlatformBinaries verifies if plugin tar contains at least the binaries expected
-func hasExpectedPlatformBinaries(filePath string, expectedBinaries map[string]string) error {
-	f, err := os.Open(filePath)
+// findPlatformBinaries finds the binaries for which the plugin was compiled
+func findPlatformBinaries(filePath string) (map[string]string, error) {
+	tmpDir, err := os.MkdirTemp("", "platform-plugin-*")
 	if err != nil {
-		return errors.Wrap(err, "failed to setup sftp client")
+		return nil, errors.Wrap(err, "failed to create temporary directory")
 	}
-	defer f.Close()
+	defer os.RemoveAll(tmpDir) // clean up
 
-	gzf, err := gzip.NewReader(f)
+	err = unpackPlugin(filePath, tmpDir)
 	if err != nil {
-		return errors.Wrap(err, "failed to read gzip data")
+		return nil, errors.Wrap(err, "failed to unpack plugin")
 	}
 
-	foundBinaries := make(map[string]bool, len(expectedBinaries))
-	for _, v := range expectedBinaries {
-		foundBinaries[v] = false
+	dir, err := os.ReadDir(tmpDir)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read tmp dir")
 	}
 
-	tarReader := tar.NewReader(gzf)
-	for {
-		header, err := tarReader.Next()
-		if err == io.EOF {
-			break
-		}
-		if err != nil {
-			return errors.Wrap(err, "failed to read tar data")
-		}
+	// If the root of the plugin bundle consists of exactly one directory, assume the plugin
+	// is contained therein. Otherwise the root directory is expected to contain the plugin.
+	if len(dir) == 1 && dir[0].IsDir() {
+		tmpDir = filepath.Join(tmpDir, dir[0].Name())
+	}
 
-		name := header.FileInfo().Name()
+	manifest, _, err := model.FindManifest(tmpDir)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to find manifest")
+	}
 
-		if header.Typeflag == tar.TypeReg {
-			foundBinaries[name] = true
-		}
+	// We should probably support this as a platform-agnostic plugin, but leaving that existing
+	// gap to a future reader.
+	if manifest.Server == nil {
+		return nil, fmt.Errorf("no server defined")
 	}
 
-	var missingBinaries []string
-	for k, v := range foundBinaries {
-		if !v {
-			missingBinaries = append(missingBinaries, k)
+	if len(manifest.Server.Executables) == 0 {
+		if len(manifest.Server.Executable) > 0 {
```

Comment:
Backwards compatibility? :exploding_head:

Reply:
Some developer, in a universe far, far away, will thank me for the clarity of this error message ;P